applications/addons/models/class.addonmodel.php

```diff
@@ -775,6 +770,56 @@ public function updateCurrentVersion($AddonID) {
             $this->SQL->history()->put('Addon', array('CurrentAddonVersionID' => $MaxVersion->Version), array('AddonID' => $AddonID));
         }
     }
+
+    /**
+     * Parse a slug and returns information array about the plugin.
+     *
+     * Possible values of the returned array are:
+     *   - version: the Version number
+     *   - type: one of self::$Types
+     *   - typeID: numeric key for Vanilla's addon types
+     *   - key: the key of the plugin
+     *
+     * @param string $slug The slug to parse.
+     *
+     * @return array Addon info (version, type, type id and key)
+     */
+    public function parseSlug($slug) {
+        $parts = explode('-', $slug);
+
+        if (is_numeric($parts[0])) {
+            // Slug is numeric ID.
+            return ['key' => $parts[0]];
+        }
+
+        $end = array_pop($parts);
+        $type = strtolower($end);
+        if (in_array($type, self::$Types)) {
+            // Slug ends with one of self::Types without version number.
+            return [
+                'version' => false,
+                'type' => $type,
+                'typeID' => self::$Types[$type],
+                'key' => implode('-', $parts)
+            ];
+        }
+
+        // If $end is no addon type, it must the version number.
+        $version = $end;
+        $end = array_pop($parts);
+        if (in_array($end, self::$Types)) {
```

Comment:
This should check for the key in `self::$Types`, not the value.
Also, this function doesn't work when downloading the addon (where the addon version is concatenated to the end of the slug).

Reply:
I have changed that.
I couldn't try the complete process, because the upload failed in my setup (even with the master branch)